data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -74,111 +78,131 @@ private TableMigrationUtil() {
+    Metrics metrics;
+    try {
+      InputFile file = HadoopInputFile.fromPath(stat.getPath(), conf);
+      long rowCount = Avro.rowCount(file);
+      metrics = new Metrics(rowCount, null, null, null, null);
+    } catch (UncheckedIOException e) {
+      throw new RuntimeException("Unable to read the footer of the avro file: " +
+              stat.getPath(), e);
+    }
+
+    String partitionKey = spec.fields().stream()
+            .map(PartitionField::name)
+            .map(name -> String.format("%s=%s", name, partitionPath.get(name)))
+            .collect(Collectors.joining("/"));
+
+    return  DataFiles.builder(spec)
+            .withPath(stat.getPath().toString())
+            .withFormat("avro")
+            .withFileSizeInBytes(stat.getLen())
+            .withMetrics(metrics)
+            .withPartitionPath(partitionKey)
+            .build();
+  }
+
+  private static DataFile getParquetDataFile(FileStatus stat, Map<String, String> partitionPath,
                                                      PartitionSpec spec, Configuration conf,
                                                      MetricsConfig metricsSpec, NameMapping mapping) {
+    Metrics metrics;
     try {
-      Path partition = new Path(partitionUri);
-      FileSystem fs = partition.getFileSystem(conf);
-
-      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
-          .filter(FileStatus::isFile)
-          .map(stat -> {
-            Metrics metrics;
-            try {
-              ParquetMetadata metadata = ParquetFileReader.readFooter(conf, stat);
-              metrics = ParquetUtil.footerMetrics(metadata, Stream.empty(), metricsSpec, mapping);
-            } catch (IOException e) {
-              throw new RuntimeException("Unable to read the footer of the parquet file: " +
-                  stat.getPath(), e);
-            }
-            String partitionKey = spec.fields().stream()
-                .map(PartitionField::name)
-                .map(name -> String.format("%s=%s", name, partitionPath.get(name)))
-                .collect(Collectors.joining("/"));
-
-            return DataFiles.builder(spec)
-                .withPath(stat.getPath().toString())
-                .withFormat("parquet")
-                .withFileSizeInBytes(stat.getLen())
-                .withMetrics(metrics)
-                .withPartitionPath(partitionKey)
-                .build();
-          }).collect(Collectors.toList());
-    } catch (IOException e) {
-      throw new RuntimeException("Unable to list files in partition: " + partitionUri, e);
+      InputFile file = HadoopInputFile.fromPath(stat.getPath(), conf);
+      metrics = ParquetUtil.fileMetrics(file, metricsSpec, mapping);
+    } catch (UncheckedIOException e) {
+      throw new RuntimeException("Unable to read the footer of the parquet file: " +
+              stat.getPath(), e);
     }
+
+    String partitionKey = spec.fields().stream()
```

Comment:
This part is also identical for each file type so we can pull this out of the format specific code branches

Reply:
Ok  ,  I also  pull out of the  `DataFiles.builder`